### ISSUE TITLE
Extremely bad interference between zstd and malloc

### ISSUE THREAD

**Issue by Rogier-5** [facebook/zstd#422]
Zstd streaming compression interferes extremely badly with the malloc library on linux.

What happens, is that on every compression, zstd will allocate memory, which causes `malloc()` to `mmap()` one 2MiB and two 5 MiB areas. After compression finishes, both areas are `munmap()ped` again. In the process, at least one of the areas is also `memset()` to zero. All of this is a huge waste of CPU cycles.

I modified the included example code to demonstrate the problem.
Apply the following patch:

``` diff
diff --git a/examples/streaming_compression.c b/examples/streaming_compression.c
index 108a63c..f089f62 100644
--- a/examples/streaming_compression.c
+++ b/examples/streaming_compression.c
@@ -123,8 +123,11 @@ int main(int argc, const char** argv)
         return 1;
     }

-    const char* const outFilename = createOutFilename_orDie(inFilename);
-    compressFile_orDie(inFilename, outFilename, 1);
+    int i;
+    for (i = 0; i < 1000; i++) {
+       const char* const outFilename = createOutFilename_orDie(inFilename);
+       compressFile_orDie(inFilename, outFilename, 7);
+    }

     return 0;
 }
```

And run the example on an input file, e.g. its own source file.
Using `time` shows that most time is spent in the OS.
Use `strace -c` to see that most time is spent in `munmap()`
(in particular: over 5000 `munmap()` calls for compressing 1000 files...)

Apply the following additional patch:

``` diff
diff --git a/examples/streaming_compression.c b/examples/streaming_compression.c
index 108a63c..717918c 100644
--- a/examples/streaming_compression.c
+++ b/examples/streaming_compression.c
@@ -13,6 +13,7 @@
 #include <errno.h>     // errno
 #define ZSTD_STATIC_LINKING_ONLY  // streaming API defined as "experimental" for the time being
 #include <zstd.h>      // presumes zstd library is installed
+#include <malloc.h>


 static void* malloc_orDie(size_t size)
@@ -116,6 +117,9 @@ int main(int argc, const char** argv)
     const char* const exeName = argv[0];
     const char* const inFilename = argv[1];

+    mallopt(M_MMAP_MAX, 0);
+    mallopt(M_TRIM_THRESHOLD, 1024*1024*100);
+
     if (argc!=2) {
         printf("wrong arguments\n");
         printf("usage:\n");
```

and observe the performance difference.

**Comment by Rogier-5** [comment]
> In this example, `compressFile_orDie()` allocates and frees memory for each compression operation. The amount of reserved memory is highly dependent on compression level, so level 7 definitely requires more than level 1.

> But if you know in advance that your application is going to compress a lot of files / messages / blocks, the recommended process is to keep `ZSTD_CStream()` alive, and re-use it for each file. It's enough to use `ZSTD_initCStream()` for that.

Actually, I did. It saves one 5MB and the 2MB `mmap()`/`munmap()`, but the remaining 5MB allocation remains, just like the `memset()`.

> The main difference is that memory will be re-used. This will dramatically reduce the load on `malloc()`.

With compression level 7, I observed a 50% reduction in compression time by keeping `ZSTD_CStream()` alive. So it makes a significant difference indeed.

However, tuning malloc to eliminate both the repeated `mmap()`, `munmap()` system calls and the `memset()`, reduced the compression time (at level 7) by another 80%! That makes the mere 50% savings from not calling `ZSTD_CStream()` almost seem insignificant...

![graph of zstd performance comparison](https://cloud.githubusercontent.com/assets/6202053/19588203/6e392dd8-9764-11e6-93fe-921410b050d2.png)

The top line is the original code. The second line eliminates the stream creation, and the third data set was made after tuning `malloc()`.

> In compressing myriad of small files, it makes a large runtime difference.

The overhead would certainly be relatively lower for large files. I don't know if the regular (non-streaming) API also exhibits this anomalous behavior when compressing smaller files. The files in my tests were mostly 16KB, so they weren't _that_ small. Given the magnitude of the performance impact (i.e. a factor 10 in running time - in my case), I would expect significantly larger files to be affected as well.

In any case, it might be useful to at least document this anomalous behavior, and the 'solution', so that people will be aware of it.

**Comment by Cyan4973** [comment]
> Actually, I did. It saves one 5MB and the 2MB `mmap()`/`munmap()`, but the remaining 5MB allocation remains, just like the `memset()`.

It's unclear what was precisely tested.
Note that `compressFile_orDie()` is not `libzstd`, it's just a simple example, not prepared to handle multiple files in a row.
An optimised implementation should not allocate anything beyond the first time.
You may have a look at `fileio.c` which implements this concept.

`memset()` remains a significant workload if applied to multi-MB memory region.
However, even this workload should disappear starting `v1.1.0`,
in scenarios compressing multiple files in a row with same compression settings.
Not all `memset()`will disappear, but the ones remaining should operate on KB-range memory regions, hence be a non-issue.
If your experience is different, please report, as it is something we want to optimize properly.

> In any case, it might be useful to at least document this anomalous behavior, and the 'solution', so that people will be aware of it.

Sure

**Comment by Cyan4973** [comment]
> In any case, it might be useful to at least document this anomalous behavior, and the 'solution', so that people will be aware of it.

[`zstd.h`](https://github.com/facebook/zstd/blob/dev/lib/zstd.h) code comments have been amended to include more context information about context re-use and its impact on memory management

**Comment by Cyan4973** [comment]
There's a new [example](https://github.com/facebook/zstd/tree/dev/examples), called [multiple_streaming_compression](https://github.com/facebook/zstd/blob/dev/examples/multiple_streaming_compression.c),
which shows how to make multiple compression operations in a row using the streaming API,
preserving memory by allocating it only once at the beginning.

**Comment by Rogier-5** [comment]
Thanks for adding the example and the additional documentation.

I reran some tests, and collected some more detailed information, using both example programs, compressing the same 16kB text file 1000 times. I changed the streaming version to compress at level 7 as well. For every test, the output of `strace -c`, and then the output of `/usr/bin/time` is below. The combined patch is at the bottom (`mallopt()` calls, and other modifications. Disable the `mallopt()` calls for the 'unmodified' tests).

Summary:

Test | Running time | system time | % system time in `munmap()` calls
---- | ------------ | ----------- | --------------------------
`streaming_compression.c` (without `mallopt()`) | 11s | 62% | 92%
`streaming_compression.c` (with `mallopt()`) | 3.3s | 11.8% | 0%
`multiple_streaming_compression.c` (without `mallopt()`) | 5.6s | 62% | 92%
`multiple_streaming_compression.c` (with `mallopt()`) | 1.3s | 12.3% | 0%

So not using `mallopt()` in `multiple_streaming_compression.c` increases the running time by a factor 4.3. Freeing the stream allocated is good for another factor 2 increase.

If `mallopt()` *is* used, then freeing the stream is relatively even more expensive, causing an increase of a factor 2.5 ! (1.3 -> 3.3)

I did the same tests on a 800Kb file, compressing it 1000 times as well using `multiple_streaming_compression`, and using `mallopt()` still saves 22% (user+system time); considering just system time, the savings were 70%. With 100 and 10 files, I got savings of 20% and 23% (user+system) respectively, with system time savings of 60% and 75%.

Raw output (1000 files of 16kB):
('unmodified' means: not using `mallopt()`. They _were_ modified to compress the same file 1000 times, and `streaming_compression` was modified to use compression level 7). See the patch below.

Unmodified version of `streaming_compression`
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 92.40    0.093654          19      5002           munmap
  4.58    0.004638           2      2002           open
  1.89    0.001911           1      2002           close
  0.41    0.000413           0      5007           mmap2
  0.31    0.000315           0      1000           write
  0.19    0.000196           0      3001           read
  0.15    0.000151           0      3005           brk
  0.07    0.000075           0      2002           fstat64
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         3         3 access
  0.00    0.000000           0         2           mprotect
  0.00    0.000000           0         1           set_thread_area
------ ----------- ----------- --------- --------- ----------------
100.00    0.101353                 23028         3 total
3.86user 6.95system 0:11.06elapsed 97%CPU (0avgtext+0avgdata 10536maxresident)k
0inputs+8000outputs (0major+2320067minor)pagefaults 0swaps
```
Version of `streaming_compression` modified to call `mallopt`:
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 95.63    0.002144           1      2002           close
  2.77    0.000062           0      2006           mmap2
  1.61    0.000036           0      2002           open
  0.00    0.000000           0      3001           read
  0.00    0.000000           0      1000           write
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         3         3 access
  0.00    0.000000           0         8           brk
  0.00    0.000000           0      2001           munmap
  0.00    0.000000           0         2           mprotect
  0.00    0.000000           0      2002           fstat64
  0.00    0.000000           0         1           set_thread_area
------ ----------- ----------- --------- --------- ----------------
100.00    0.002242                 14029         3 total
2.84user 0.38system 0:03.32elapsed 97%CPU (0avgtext+0avgdata 10600maxresident)k
0inputs+8000outputs (0major+3387minor)pagefaults 0swaps
compressed 1 files 
```
Unmodified version of `multiple_streaming_compression`:
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 91.77    0.032291          11      3006           munmap
  3.34    0.001174           1      2002           close
  2.48    0.000874           0      2002           open
  0.68    0.000238           0      3012           mmap2
  0.67    0.000236           0      3001           read
  0.62    0.000219           0      1001           write
  0.44    0.000154           0      2003           fstat64
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         3         3 access
  0.00    0.000000           0         3           brk
  0.00    0.000000           0         2           mprotect
  0.00    0.000000           0         1           set_thread_area
------ ----------- ----------- --------- --------- ----------------
100.00    0.035186                 16037         3 total
compressed 1 files 
2.06user 3.45system 0:05.63elapsed 97%CPU (0avgtext+0avgdata 10596maxresident)k
0inputs+8000outputs (0major+1155234minor)pagefaults 0swaps
compressed 1 files 
```
Version of `multiple_streaming_compression` modified to call `mallopt`:
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 31.74    0.000233           0      3001           read
 28.07    0.000206           0      2002           open
 23.16    0.000170           0      2002           close
 11.31    0.000083           0      2003           fstat64
  5.72    0.000042           0      1001           write
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         3         3 access
  0.00    0.000000           0         8           brk
  0.00    0.000000           0      2001           munmap
  0.00    0.000000           0         2           mprotect
  0.00    0.000000           0      2007           mmap2
  0.00    0.000000           0         1           set_thread_area
------ ----------- ----------- --------- --------- ----------------
100.00    0.000734                 14032         3 total
compressed 1 files 
1.07user 0.15system 0:01.27elapsed 96%CPU (0avgtext+0avgdata 10652maxresident)k
0inputs+8000outputs (0major+3387minor)pagefaults 0swaps
```
Patch (for repeating my non-`mallopt()` tests, comment out the `mallopt()` calls):
```diff
diff --git a/examples/multiple_streaming_compression.c b/examples/multiple_streaming_compression.c
index c002a88..5bf2ece 100644
--- a/examples/multiple_streaming_compression.c
+++ b/examples/multiple_streaming_compression.c
@@ -16,6 +16,7 @@
 #include <stdio.h>     // fprintf, perror, feof
 #include <string.h>    // strerror
 #include <errno.h>     // errno
+#include <malloc.h>
 #define ZSTD_STATIC_LINKING_ONLY  // streaming API defined as "experimental" for the time being
 #include <zstd.h>      // presumes zstd library is installed
 
@@ -126,6 +127,9 @@ static void compressFile_orDie(resources ress, const char* fname, const char* ou
 
 int main(int argc, const char** argv)
 {
+    mallopt(M_MMAP_MAX, 0);
+    mallopt(M_TRIM_THRESHOLD, 1024*1024*100);
+
     const char* const exeName = argv[0];
 
     if (argc<2) {
@@ -139,8 +143,9 @@ int main(int argc, const char** argv)
     void* ofnBuffer = NULL;
     size_t ofnbSize = 0;
 
-    int argNb;
-    for (argNb = 1; argNb < argc; argNb++) {
+    int i;
+    int argNb = 1;
+    for (i = 0; i < 1000; i++) {
         const char* const ifn = argv[argNb];
         size_t const ifnSize = strlen(ifn);
         size_t const ofnSize = ifnSize + 5;
diff --git a/examples/streaming_compression.c b/examples/streaming_compression.c
index 108a63c..26b43f4 100644
--- a/examples/streaming_compression.c
+++ b/examples/streaming_compression.c
@@ -7,6 +7,7 @@
  */
 
 
+#include <malloc.h>
 #include <stdlib.h>    // malloc, exit
 #include <stdio.h>     // fprintf, perror, feof
 #include <string.h>    // strerror
@@ -113,6 +114,9 @@ static const char* createOutFilename_orDie(const char* filename)
 
 int main(int argc, const char** argv)
 {
+    mallopt(M_MMAP_MAX, 0);
+    mallopt(M_TRIM_THRESHOLD, 1024*1024*100);
+
     const char* const exeName = argv[0];
     const char* const inFilename = argv[1];
 
@@ -124,7 +128,9 @@ int main(int argc, const char** argv)
     }
 
     const char* const outFilename = createOutFilename_orDie(inFilename);
-    compressFile_orDie(inFilename, outFilename, 1);
+    int i;
+    for (i=0; i<1000; i++)
+    compressFile_orDie(inFilename, outFilename, 7);
 
     return 0;
 }
```

**Comment by Rogier-5** [comment]
I did a quick tests using `programs/zstd`, and it does not do excessive `munmap()` calls, even when compressing the same file 1000 times. It seems the major difference is that it calls `ZSTD_initCStream_advanced()` only once, instead of once for every file. Is there a reason why `ZSTD_initCStream()` must be called once for every compression (like it is in `multiple_streaming_compression.c`), while `ZSTD_initCStream_advanced()` can be called just once for may compressions, like in `fileio.c` ?

Would it be that the effects of both only depend on the parameters used, and it's really only necessary to call `ZSTD_initCStream()` again if the parameters (i.e. the compression level) change ?

**Comment by Cyan4973** [comment]
Thanks very much for the very interesting investigation @Rogier-5 .
It's a bit different baseline expectation, which proves there is something to look for.
I'm going to review deeper into it.

**Comment by Cyan4973** [comment]
quick suggestion : could it be related to `fopen`/`fclose` ?
I note these 2 calls are quite larger without `mallopt`.

In which case, a difference with the example happens when using `zstd` in combination with `stdout` (`-c`) : the output is only opened and closed once, instead of once per file.

**Comment by Cyan4973** [comment]
> It seems the major difference is that it calls ZSTD_initCStream_advanced() only once, instead of once for every file. 

`fileio.c` uses `ZSTD_resetCStream()`, which, in essence, is supposed to be the same as `ZSTD_initCStream()` but with the implicit order to re-use the same compression parameters as previous operation.

Both should be technically equivalent when compressing repetitively the same file with the same compression level. But let's look deeper into it, to be sure.

**Comment by Cyan4973** [comment]
OK, got it.
Inside `CStream` there is a digested dictionary structure `CDict` which is allocated and init every time `ZSTD_initCStream()` is invoked (but which is skipped when using `ZSTD_resetCStream()`), even when there is no dictionary.

Getting rid of it is more complex than it sounds, due to a chain of dependencies, but this will have to be done in order to remove this useless allocation effort.

**Comment by Rogier-5** [comment]
> Getting rid of it is more complex than it sounds, due to a chain of dependencies, but this will have to be done in order to remove this useless allocation effort.

From the behavior, it wouldn't have to disappear. By default, a static buffer could be used instead for example (I assume it's read-only ?). The static version would be replaced with a dynamic version only when needed. Or else, it seems that it may also be sufficient if the buffer, once allocated, is not deallocated (as apparently it's `munmap()` and not `mmap()` that consumes most CPU time). The previous buffer could be kept, stored inside the stream object, and reused in the next `ZSTD_initCStream()`. It would only be deallocated once the stream is freed.

Does that sound sensible ? It might save some effort. Or am I misunderstanding the issue ?

**Comment by Rogier-5** [comment]
BTW: given this behavior when freeing the stream (which is apparently expensive, regardless of the presence of a custom dict - assuming it's indeed munmap that causes the problem), the recommended procedure would have to be to avoid freeing the stream unless absolutely necessary. A progam exiting without cleanup would be much more efficient than a program that carefully cleans up before exiting...

**Comment by Cyan4973** [comment]
Latest update in "dev" branch should eliminate the observed `mmap`/`munmap` behavior, by keeping `cdict` into `NULL` state when no dictionary is used.

**Comment by Cyan4973** [comment]
Fix is integrated into v1.1.1
